### ISSUE TITLE
[Effect] Mag Point + Min Anisotropic not available

### ISSUE THREAD

**Issue by mklingen** [FNA-XNA/FNA#70]
I've noticed that FNA assumes that any texture sampler that uses Anisotropic min or mag filters must mean that the desired texture sampler is "Anisotropic " -- but my shaders use Point for magnification and Anisotropic for minification. This might seem like a bizarre case, but it is the filtering that looks nicest for pixel art.

**Comment by flibitijibibo** [comment]
Can you make a small repro case for this? At the time I wasn't sure if those two states were even compatible with one another in XNA, so if we can see how they do it in D3D9 for XNA, we should be able to roughly approximate it in OpenGL. 99% of how we approximate it is in XNAToGL.MagFilter and MinMipFilter so it shouldn't be _too_ hard to rework this internally; the main problem is representing it in the SamplerState object, which definitely did not account for such an occasion. I wonder what the SamplerState says after that shader is applied!

**Comment by mklingen** [comment]
Ethan, here is a simple example in XNA. https://github.com/mklingen/XNAAnisotropicExample

I have been looking at the SamplerState, which is really bizarre. SamplerState[0] is null, the rest are LinearWrap.

**Comment by flibitijibibo** [comment]
We'll have to look with apitrace then... we can still get the sampler state we want in the GLDevice but it will be wrong in the C# portion.

**Comment by flibitijibibo** [comment]
Here's a possible fix:

```
diff --git a/src/Graphics/Effect/Effect.cs b/src/Graphics/Effect/Effect.cs
index eb2951e..86dd18c 100644
--- a/src/Graphics/Effect/Effect.cs
+++ b/src/Graphics/Effect/Effect.cs
@@ -897,14 +897,7 @@ namespace Microsoft.Xna.Framework.Graphics
 				}
 				if (filterChanged)
 				{
-					if (	magFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_ANISOTROPIC ||
-						minFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_ANISOTROPIC ||
-						mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_ANISOTROPIC	)
-					{
-						// Just assume we wanted Anisotropic if any of these qualify.
-						filter = TextureFilter.Anisotropic;
-					}
-					else if (magFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_POINT)
+					if (magFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_POINT)
 					{
 						if (minFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_POINT)
 						{
@@ -913,7 +906,8 @@ namespace Microsoft.Xna.Framework.Graphics
 							{
 								filter = TextureFilter.Point;
 							}
-							else if (mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR)
+							else if (	mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR ||
+									mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_ANISOTROPIC	)
 							{
 								filter = TextureFilter.PointMipLinear;
 							}
@@ -922,14 +916,16 @@ namespace Microsoft.Xna.Framework.Graphics
 								throw new NotImplementedException("Unhandled mipfilter type!");
 							}
 						}
-						else if (minFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR)
+						else if (	minFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR ||
+								minFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_ANISOTROPIC	)
 						{
 							if (	mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_NONE ||
 								mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_POINT	)
 							{
 								filter = TextureFilter.MinLinearMagPointMipPoint;
 							}
-							else if (mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR)
+							else if (	mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR ||
+									mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_ANISOTROPIC	)
 							{
 								filter = TextureFilter.MinLinearMagPointMipLinear;
 							}
@@ -943,7 +939,8 @@ namespace Microsoft.Xna.Framework.Graphics
 							throw new NotImplementedException("Unhandled minfilter type!");
 						}
 					}
-					else if (magFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR)
+					else if (	magFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR ||
+							magFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_ANISOTROPIC	)
 					{
 						if (minFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_POINT)
 						{
@@ -952,7 +949,8 @@ namespace Microsoft.Xna.Framework.Graphics
 							{
 								filter = TextureFilter.MinPointMagLinearMipPoint;
 							}
-							else if (mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR)
+							else if (	mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR ||
+									mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_ANISOTROPIC	)
 							{
 								filter = TextureFilter.MinPointMagLinearMipLinear;
 							}
@@ -961,14 +959,16 @@ namespace Microsoft.Xna.Framework.Graphics
 								throw new NotImplementedException("Unhandled mipfilter type!");
 							}
 						}
-						else if (minFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR)
+						else if (	minFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR ||
+								minFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_ANISOTROPIC	)
 						{
 							if (	mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_NONE ||
 								mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_POINT	)
 							{
 								filter = TextureFilter.LinearMipPoint;
 							}
-							else if (mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR)
+							else if (	mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_LINEAR ||
+									mipFilter == MojoShader.MOJOSHADER_textureFilterType.MOJOSHADER_TEXTUREFILTER_ANISOTROPIC	)
 							{
 								filter = TextureFilter.Linear;
 							}
```

Since Anisotropic seems to confuse XNA4 we're going to take the closest accurate path on the GL side and just treat that state as Linear, letting MaxAnisotropy do the rest of the work. Does this produce closer results?

**Comment by flibitijibibo** [comment]
Should have mentioned: The results are definitely different and the GL is at least putting GL_LINEAR and GL_NEAREST in the right places, but I dunno if it's visually accurate to what you want.

**Comment by flibitijibibo** [comment]
Patch pushed to repo since it's at least more accurate with regard to point/linear filters, all that's left is to figure out if MaxAnisotropy has to be touched at all: https://github.com/FNA-XNA/FNA/commit/9578e028bdfa62b5cb38e91c55d10be2a4de9f5b

**Comment by flibitijibibo** [comment]
Closing due to old age, though as far as I know this is fixed in FNA 16.12 and newer. Let me know if this isn't the case.
